filament/backend/src/webgpu/platform/WebGPUPlatform.cpp

```diff
@@ -67,72 +75,254 @@ namespace {
+[[nodiscard]] std::unordered_set<AdapterDetails, AdapterDetailsHash> requestUniqueAdapters(
+        wgpu::Instance const& instance, wgpu::Surface const& surface) {
+    // make a series of requests asynchronously for different combinations applicable to the
+    // platform...
+    std::array<wgpu::RequestAdapterOptions, WebGPUPlatform::MAX_ADAPTER_REQUESTS> requests{};
+    size_t requestCount = 0;
+    WebGPUPlatform::getAdapterOptions(requests, requestCount);
+    std::unordered_set<AdapterDetails, AdapterDetailsHash> uniqueAdapters;
+    uniqueAdapters.reserve(requestCount);
+    std::mutex adaptersMutex;
+    std::array<wgpu::Future, requests.size()> futures{};
+    for (size_t i = 0; i < requestCount; i++) {
+        wgpu::RequestAdapterOptions& options = requests[i];
+        options.compatibleSurface = surface;
+        futures[i] = instance.RequestAdapter(&options, wgpu::CallbackMode::WaitAnyOnly,
+                [&options, &uniqueAdapters, &adaptersMutex](wgpu::RequestAdapterStatus const status,
+                        wgpu::Adapter const& readyAdapter, wgpu::StringView const message) {
+                    FILAMENT_CHECK_POSTCONDITION(
+                            status != wgpu::RequestAdapterStatus::CallbackCancelled)
+                            << "Failed to request a WebGPU adapter due to the request callback "
+                               "being cancelled? Options: "
+                            << toString(options) << " " << message.data;
+                    FILAMENT_CHECK_POSTCONDITION(status != wgpu::RequestAdapterStatus::Error)
+                            << "Failed to request a WebGPU adapter due to an error. Options: "
+                            << toString(options) << " Error: " << message.data;
+                    if (status == wgpu::RequestAdapterStatus::Success) {
+                        AdapterDetails details = AdapterDetails(readyAdapter);
+                        FILAMENT_CHECK_POSTCONDITION(readyAdapter.GetInfo(&details.info))
+                                << "Failed to get info for adapter (options: " << toString(options)
+                                << ")";
+                        const std::lock_guard<std::mutex> lock(adaptersMutex);
+                        uniqueAdapters.emplace(std::move(details.info), details.powerPreference,
+                                std::move(details.adapter));
+                        return;
+                    }
+                    assert_invariant(status == wgpu::RequestAdapterStatus::Unavailable);
+                });
+    }
+    // wait for all the results to return...
+    for (size_t i = 0; i < requestCount; i++) {
+        wgpu::RequestAdapterOptions& options = requests[i];
+        wgpu::Future& future = futures[i];
+        // TODO define reasonable timeout (and/or do this more asynchronously)
+        wgpu::WaitStatus status = instance.WaitAny(future, UINT16_MAX);
```

Comment:
UINT16_MAX is a long time to wait.  I would simply leave this at 1000 with the TODO still in place.